torch/distributed/pipelining/schedules.py

```diff
@@ -1124,12 +1132,9 @@ def __init__(
         self.pp_group_size = stages[0].group_size
         self.rank = stages[0].group_rank
         # Set the pipeline stage states
-        if stage_index_to_group_rank is None:
-            self.stage_index_to_group_rank = generate_stage_to_rank_mapping(
-                self.pp_group_size, self._num_stages
-            )
-        else:
-            self.stage_index_to_group_rank = stage_index_to_group_rank
+        self.stage_index_to_group_rank = generate_stage_to_rank_mapping(
```

Comment:
i'm confused. are we first generating via generate_... and then overwriting inside _validate by computing the same thing a different way?

Reply:
Yeah, `_validate` is used when the schedule is loaded by CSV. There are two cases:

1. Using a default schedule (`Interleaved1F1B`, `BFS`, etc.)
- During initialize, `generate_stage_to_rank_mapping`
- pipeline actions are created based on the schedule implementation

2. Creating schedule to load with CSV
- During initialize, `generate_stage_to_rank_mapping` (this is not needed and eventually gets overridden)
- load csv to get the pipeline actions
- _validate which infers the `stage_to_rank_mapping` and sets it

I think to do this in a cleaner way I will need to probably need to make `PipelineScheduleMulti` an ABC and only allow the csv loading via the runtime class?